File: 0200_NumberOfIslands.cpp
struct Point {
    int r;
    int c;
};

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        // search strategy: right -> bottom -> left -> top
        int R = grid.size();
        if (grid.size() == 0) return 0;
        int C = grid[0].size();
        
        stack<struct Point> s; // push which node to visit next
        
        int num = 0;
        
        vector<vector<bool>> visited(R, vector<bool>(C));
        struct Point p;
        
        
        for (int r = 0; r < R; r++) {
            for (int c = 0; c < C; c++) {
                if(!visited[r][c] && grid[r][c] == '1'){
                    p.r = r;
                    p.c = c;
                    s.push(p);
                    num++;
                } else continue;
                // Depth-First-Search
                while (s.size() > 0) {
                    struct Point p = s.top();
                    s.pop();
                    visited[p.r][p.c] = true;
                    struct Point q;
                    if (p.c < C-1) {
                        if (!visited[p.r][p.c+1] && grid[p.r][p.c+1] == '1') {
                            q.r = p.r;
                            q.c = p.c+1;
                            visited[q.r][q.c] = true;
                            s.push(q);
                        }
                    }
                    if (p.r < R-1) {
                        if (!visited[p.r+1][p.c] && grid[p.r+1][p.c] == '1') {
                            q.r = p.r+1;
                            q.c = p.c;
                            visited[q.r][q.c] = true;
                            s.push(q);
                        }
                    }
                    if (p.c > 0) {
                        if (!visited[p.r][p.c-1] && grid[p.r][p.c-1] == '1') {
                            q.r = p.r;
                            q.c = p.c-1;
                            visited[q.r][q.c] = true;
                            s.push(q);
                        }
                    }
                    if (p.r > 0) {
                        if(!visited[p.r-1][p.c] && grid[p.r-1][p.c] == '1') {
                            q.r = p.r-1;
                            q.c = p.c;
                            visited[q.r][q.c] = true;
                            s.push(q);
                        }
                    }
                }
            }
        }
        return num;
    }
};

File: 0200_NumberOfIslands.py
class Solution:
    def __init__(self):
        self.visited = set()
        
    def bfs(self, grid, i, j):
        R = self.R; C = self.C
        q = [(i,j)]
        while len(q) > 0:
            r, c = q.pop(0)
            if r < R-1:
                if grid[r+1][c] == "1" and (r+1,c) not in self.visited:
                    q.append((r+1,c))
                    self.visited.add((r+1,c))
            if c < C-1:
                if grid[r][c+1] == "1" and (r,c+1) not in self.visited:
                    q.append((r,c+1))
                    self.visited.add((r,c+1))
            if r > 0:
                if grid[r-1][c] == "1" and (r-1,c) not in self.visited:
                    q.append((r-1,c))
                    self.visited.add((r-1,c))
            if c > 0:
                if grid[r][c-1] == "1" and (r,c-1) not in self.visited:
                    q.append((r,c-1))
                    self.visited.add((r,c-1))
            
    
    def numIslands(self, grid: List[List[str]]) -> int:
        self.R = len(grid)
        if self.R == 0:
            return 0
        self.C = len(grid[0])
        if self.C == 0:
            return 0
        num = 0
        for i,row in enumerate(grid):
            for j,elem in enumerate(row):
                if elem == "1" and (i,j) not in self.visited:
                    self.visited.add((i,j))
                    num += 1
                    self.bfs(grid,i,j)
        return num
